### ISSUE TITLE
Installing a library doesn't install its dependencies

### ISSUE THREAD

**Issue by goldfirere** [haskell/cabal#6263]
**Describe the bug**
When I install (globally) a library, I don't seem to get its dependencies in my environment file.

**To Reproduce**

```
$ cabal v2-install --lib singletons
$ ghci
> import Language.Haskell.TH.Desugar
```

Witness an error that the requested module is in a hidden package.

**Expected behavior**
I expected to be able to access modules from libraries. It's very awkward otherwise. Running `cabal v2-install ... --only-dependencies` doesn't help.

**System informataion**
 - Mac
 - `cabal 3.0.0.0`
 - `ghc 8.8.1`

**Additional context**
Maybe I'm misusing the new Cabal, but I don't know how to install a library's set of (transitive) dependencies in a global environment file.

**Comment by phadej** [comment]
IMHO, `cabal install --lib` is simply under-developed and not ready for prime use, use `cabal repl` when possible, or tools like https://github.com/phadej/cabal-env (disclaimer: i'm the author, so obviously it works better and "right").

There are already two lessons:
- you really want to *manage* libraries, i.e. learn from `apt` about e.g. installed vs. auto-installed
- installing libraries and installing executables is so different that single command (`install`) is bad UX.
    - I'd propose `cabal env` (obviously! for managing *environments*) and `cabal install` (for installing executables), but I doubt that would be accepted.

---

For your particular use case

```
% cabal repl -b singletons
Resolving dependencies...
Build profile: -w ghc-8.6.5 -O1
In order, the following will be built (use -v for more details):
 - th-expand-syns-0.4.5.0 (lib) (requires download & build)
 - th-reify-many-0.1.9 (lib) (requires build)
 - th-orphans-0.13.9 (lib) (requires download & build)
 - th-desugar-1.9 (lib) (requires download & build)
 - singletons-2.5.1 (lib:singletons) (requires download & build)
 - fake-package-0 (lib) (first run)
Downloading  th-expand-syns-0.4.5.0
...
Starting     singletons-2.5.1 (all, legacy fallback)
Building     singletons-2.5.1 (all, legacy fallback)
Installing   singletons-2.5.1 (all, legacy fallback)
Completed    singletons-2.5.1 (all, legacy fallback)
Configuring library for fake-package-0..
Preprocessing library for fake-package-0..
Warning: No exposed modules
GHCi, version 8.6.5: http://www.haskell.org/ghc/  :? for help
Loaded GHCi configuration from /tmp/cabal-repl.-25256/setcwd.ghci
Prelude> import Language.Haskell.TH.Desugar
Prelude Language.Haskell.TH.Desugar>

**Comment by hvr** [comment]
> When I install (globally) a library, I don't seem to get its dependencies in my environment file.

btw, this isn't always a good thing; I'm also quite annoyed that `cabal repl` brings rev-deps into scope by default (and I keep to have to type this annoying cumbersome `--no-transitive-deps` flag... which lacks a short-option)

**Comment by goldfirere** [comment]
Thanks for the suggestions.

What always works against me with the new `cabal` is that I tend not to work on projects. I tend to do little experiments. When I start a GHCi session, I often don't yet know what libraries I will want by the end of it. So I like to install a large set of libraries globally, giving me the freedom to experiment without losing my state to switch environments. This was very easy in the old `cabal`. It seems nigh impossible now.

`cabal-env` looks like a possibility. But if I understand directly, it would require putting a large call to `cabal-env` in my `.bash_profile`. Is it snappy enough for that usage?

`cabal repl` doesn't really work for me, because it requires that I list every package I might want. You might then say I should make a dummy package that lists all the others... but then wouldn't I have to be in that project's directory in order to get it to work? I just want to be able to say `ghci` and do fun things.

Am I wrong here? I've been a staunch supporter of cabal, but I'm beginning to think it's time to shop elsewhere...

**Comment by phadej** [comment]
you don't need to call `cabal-env` in `.bash_profile`, the "state" is persisted in the environment file.

You can `cabal-env pkg1` today and `cabal-env pkg2` tomorrow (and `pkg1` will still be in the environment) until eventually it will complain about trying to enter ring nine of a dependency hell (then you'll need to start over, or a new environment).

TL;DR the idea is that `cabal-env` = `cabal install --lib`, but "better". For now, fixing bugs and improving UX in `cabal-env` is faster, as it's independent of `cabal-install`. There are few issues I want to fix before putting it to Hackage, but then it should be also simple to install (as `cabal install` for executables works well).

**Comment by goldfirere** [comment]
Ah -- I misunderstood `cabal-env`. Yes, that does sound like my new friend. I shall check it out.
